File: utils/detect_and_draw_bound.py
import os
import cv2
import numpy as np


debug_opencv = 0  # set to 1 for saving opencv images for debugging purposes


def detect_and_draw_bound(frame, width=640, height=480):
    global bounding_box, dx, dy
    
    # Initialize bounding box coordinates with default values
    x, y, w, h = 0, 0, 0, 0
    
    # Reshape mujoco frame [height x width, 1] to 3D array for opencv input [height, width, 3]
    frame_reshaped = frame.reshape((height, width, 3))

    # convert frame to BGR - I think it's already coming in as BGR for some reason
    frame_bgr = cv2.cvtColor(frame_reshaped, cv2.COLOR_RGB2BGR)

    # Convert frame to HSV (Hue, Saturation, Value) color space for easier color detection
    hsv_frame = cv2.cvtColor(frame_bgr, cv2.COLOR_BGR2HSV)

    # Define range for color in HSV (link helps define range of values for colors)
    # https://stackoverflow.com/questions/47483951/how-can-i-define-a-threshold-value-to-detect-only-green-colour-objects-in-an-ima/47483966#47483966
    lower_color = np.array([20, 0, 0])  # yellow lower
    upper_color = np.array([35, 255, 255])  # yellow upper

    # Create a mask for detecting color objects in the frame
    color_mask = cv2.inRange(hsv_frame, lower_color, upper_color)

    # Find contours in the mask
    contours, _ = cv2.findContours(color_mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)

    for contour in contours:
        # Filter out small contours that may be noise
        if cv2.contourArea(contour) > 500:
            # Get bounding box for each yellow object
            x, y, w, h = cv2.boundingRect(contour)

            # Draw a green bounding box around the yellow object
            cv2.rectangle(frame_bgr, (x, y), (x+w, y+h), (0, 255, 0), 2)       
        
    frame_bdbox = cv2.cvtColor(frame_bgr, cv2.COLOR_BGR2RGB)
    frame_boundbox = frame_bdbox.flatten()
    
    bounding_box = np.array([x, y, x+w, y+h])
    
    # center of the bounding box
    bb_center_x = (bounding_box[0] + bounding_box[2]) / 2
    bb_center_y = (bounding_box[1] + bounding_box[3]) / 2

    # center of the frame
    frame_center_x = width / 2
    frame_center_y = height / 2

    # difference between the centers
    dx = frame_center_x - bb_center_x
    dy = frame_center_y - bb_center_y
    
    if debug_opencv:
        dirname = os.path.dirname(os.path.abspath(__file__))

        # Get the parent directory of the current script's directory
        parent_dir = os.path.dirname(dirname)
        
        filename = os.path.join(parent_dir, 'images/test_hsvframe.png')
        cv2.imwrite(filename, cv2.flip(hsv_frame, -1))

        filename = os.path.join(parent_dir, 'images/test_bgr_filtered.png')
        cv2.imwrite(filename, cv2.flip(frame_bgr, -1))
        
        filename = os.path.join(parent_dir, 'images/test_filtered.png')
        cv2.imwrite(filename, cv2.flip(frame_bdbox, -1))
    
    return frame_boundbox, bounding_box, dx, dy

File: utils/detect_and_draw_bound.cpp
#include <opencv2/opencv.hpp>
#include <array>
#include <iostream>

#include "boundingbox.hpp"

bool debug_opencv = false;


void detect_and_draw_bound(BoundingBox& result) {

    cv::Mat rgbFrame = result.rgbFrame;
    cv::Mat depth8Frame = result.depth8Frame;
    int height = result.frame_height;
    int width  = result.frame_width;

    // Convert RGB to BGR - incoming is in bgr, but won't display bounding box if it isn't "converted" to bgr
    cv::Mat bgrFrame;
    cv::cvtColor(rgbFrame, bgrFrame, cv::COLOR_RGB2BGR);

    // Convert BGR to HSV (Hue, Saturation, Value) color space for easier color detection
    cv::Mat hsvFrame;
    cv::cvtColor(bgrFrame, hsvFrame, cv::COLOR_BGR2HSV);

    // Define range for color in HSV (link helps define range of values for colors)
    // https://stackoverflow.com/questions/47483951/how-can-i-define-a-threshold-value-to-detect-only-green-colour-objects-in-an-ima/47483966#47483966
    cv::Scalar lowerColor(20, 0, 0);  // color lower
    cv::Scalar upperColor(30, 255, 255); // color upper
    cv::Mat colorMask;
    cv::inRange(hsvFrame, lowerColor, upperColor, colorMask);

    // Find contours
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(colorMask, contours, cv::RETR_TREE, cv::CHAIN_APPROX_SIMPLE);

    // Iterate over contours and draw bounding boxes
    cv::Rect maxRect;
    for (const auto& contour : contours) {
        if (cv::contourArea(contour) > 500) { // Filter small contours that may be noise
            cv::Rect rect = cv::boundingRect(contour);
            if (rect.area() > maxRect.area()) {
                maxRect = rect;

                // Draw the bounding box on the original frame
                cv::rectangle(rgbFrame, maxRect, cv::Scalar(0, 255, 0), 2);
            }
        }
    }
    
    // cv::cvtColor(rgbFrame, rgbFrame, cv::COLOR_BGR2RGB);
    result.rgbFrame = rgbFrame;

    result.boundingBox = {maxRect.x, maxRect.y, maxRect.x + maxRect.width, maxRect.y + maxRect.height};
    
    // Calculate center of bounding box and frame
    result.dx = (width - (result.boundingBox[0] + result.boundingBox[2])) / 2.0;
    result.dy = (height - (result.boundingBox[1] + result.boundingBox[3])) / 2.0;

    // Debugging: Save images for verification
    if (debug_opencv) {
        cv::Mat rgb_flipped;
        cv::flip(rgbFrame, rgb_flipped, -1);
        cv::imwrite("images/frame_rgb.png", rgb_flipped);

        cv::Mat bgr_flipped;
        cv::flip(bgrFrame, bgr_flipped, -1);
        cv::imwrite("images/frame_bgr.png", bgr_flipped);

        cv::Mat hsv_flipped;
        cv::flip(hsvFrame, hsv_flipped, -1);
        cv::imwrite("images/frame_hsv.png", hsv_flipped);

        cv::Mat result_flipped;
        cv::flip(result.rgbFrame, result_flipped, -1);
        cv::imwrite("images/frame_boundbox.png", result_flipped);
    }
}
